nexus/tests/integration_tests/quotas.rs

```diff
@@ -0,0 +1,280 @@
+        &self,
+        client: &ClientTestContext,
+        name: &str,
+        size: u32,
+    ) -> Result<TestResponse, Error> {
+        NexusRequest::objects_post(
+            client,
+            "/v1/disks?project=project",
+            &params::DiskCreate {
+                identity: IdentityMetadataCreateParams {
+                    name: name.parse().unwrap(),
+                    description: "".into(),
+                },
+                size: ByteCount::from_gibibytes_u32(size),
+                disk_source: params::DiskSource::Blank {
+                    block_size: params::BlockSize::try_from(512).unwrap(),
+                },
+            },
+        )
+        .authn_as(self.auth.clone())
+        .execute()
+        .await
+    }
+}
+
+async fn setup_silo_with_quota(
+    client: &ClientTestContext,
+    silo_name: &str,
+    quotas: params::SiloQuotasCreate,
+) -> ResourceAllocator {
+    let silo = object_create(
+        client,
+        "/v1/system/silos",
+        &params::SiloCreate {
+            identity: IdentityMetadataCreateParams {
+                name: silo_name.parse().unwrap(),
+                description: "".into(),
+            },
+            quotas,
+            discoverable: true,
+            identity_mode: shared::SiloIdentityMode::LocalOnly,
+            admin_group_name: None,
+            tls_certificates: vec![],
+            mapped_fleet_roles: Default::default(),
+        },
+    )
+    .await;
+
+    populate_ip_pool(&client, "default", None).await;
+
+    // Create a silo user
+    let user = create_local_user(
+        client,
+        &silo,
+        &"user".parse().unwrap(),
+        params::UserPassword::LoginDisallowed,
+    )
+    .await;
+
+    // Make silo admin
+    grant_iam(
+        client,
+        format!("/v1/system/silos/{}", silo_name).as_str(),
+        SiloRole::Admin,
+        user.id,
+        AuthnMode::PrivilegedUser,
+    )
+    .await;
+
+    let auth_mode = AuthnMode::SiloUser(user.id);
+
+    NexusRequest::objects_post(
+        client,
+        "/v1/projects",
+        &params::ProjectCreate {
+            identity: IdentityMetadataCreateParams {
+                name: "project".parse().unwrap(),
+                description: "".into(),
+            },
+        },
+    )
+    .authn_as(auth_mode.clone())
+    .execute()
+    .await
+    .unwrap();
+
+    ResourceAllocator::new(auth_mode)
+}
+
+#[nexus_test]
+async fn test_quotas(cptestctx: &ControlPlaneTestContext) {
+    let client = &cptestctx.external_client;
+
+    // Simulate space for disks
+    DiskTest::new(&cptestctx).await;
+
+    let system = setup_silo_with_quota(
+        &client,
+        "quota-test-silo",
+        params::SiloQuotasCreate::empty(),
+    )
+    .await;
+
+    // Ensure trying to provision an instance with empty quotas fails
+    system
+        .provision_instance(client, "instance", 1, 1)
+        .await
+        .expect_err("should've failed with insufficient CPU quota");
```

Comment:
tiiiiiiny nitpick.

Whenever you do `expect_err`, we just expect **any** error, not necessarily the one you expected.

If we updated this to:

```rust
let err = system.provision_instance(...).await.expect_err("Should have failed!").to_string();
assert!(err.contains("Insufficient CPU quota"), "Unexpected error: {err}");
```

The test would be a little safer -- it would (correctly!) fail if a different error was getting spit out here

Reply:
Thanks for the reminder. I had this in my notes from our pairing session but forgot to go back and add it. Doing that now.